### PR TITLE
Add dev_build and dev_mode flags to SCons compilation instructions for desktop editor pages

### LINKED ISSUE #10018: Compiling for Windows SCons Documentation Doesn't Mention dev_build and dev_mode Flags
**Engine source version:** Godot 4.4 Dev.
**Documentation source:** Godot 4.3 English

**Issue Description:**
The instructions to setup the godot engine using SCons on the contribution section of the godot docs doesn't mention dev_build=yes or dev_mode=yes. Not using these flags can cause builds and debuggers to not load symbols and breakpoints in many code files, making it difficult to both dig through and test changes to code.

image

**URL to the documentation page:**
https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_windows.html#running-scons

### PR DESCRIPTION
**Pull request by jonathansekela**
Closes #10018. _Bugsquad edit: link to the issue so it auto-closes_

Add instructions to compile SCons with dev_build=yes and dev_mode=yes flags to desktop documentation pages.

### CODE REVIEW

File: contributing/development/compiling/compiling_for_linuxbsd.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_linuxbsd.rst:251]:
```suggestion
If you wish to add and use breakpoints while developing, add ``dev_build=yes``
```

File: contributing/development/compiling/compiling_for_linuxbsd.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_linuxbsd.rst:252]:
```suggestion
and ``dev_mode=yes`` to the command::
```

File: contributing/development/compiling/compiling_for_macos.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_macos.rst:62]:
```suggestion
If you wish to add and use breakpoints while developing, add ``dev_build=yes``
```

File: contributing/development/compiling/compiling_for_macos.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_macos.rst:63]:
```suggestion
and ``dev_mode=yes`` to the command::
```

File: contributing/development/compiling/compiling_for_windows.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_windows.rst:149]:
```suggestion
If you wish to add and use breakpoints while developing, add ``dev_build=yes``
```

File: contributing/development/compiling/compiling_for_windows.rst
Comment (AThousandShips) [inline contributing/development/compiling/compiling_for_windows.rst:150]:
```suggestion
and ``dev_mode=yes`` to the command:
```

File: contributing/development/compiling/compiling_for_linuxbsd.rst
Comment (bruvzg) [reply contributing/development/compiling/compiling_for_linuxbsd.rst:251]: `dev_build` enable additional runtime checks and disables optimization, it will make debugging more precise, but is not strictly required to debug/set brakepoints (with `debug_symbols=yes` it can be done in optimized build as well).

`dev_mode` is not related to debugging at all, it's enabling extra check during engine build.

**Comment by jonathansekela** [comment]
For @bruvzg 's mention on The dev_build parameter being the specific one making sure the debug symbols are included and dev_mode doing something slightly different, I'm assuming I should modify the documentation to reflect what each parameter does specifically?

**Comment by jonathansekela** [comment]
Alternatively, I could just put a link to the Introduction page's development and production aliases headline: https://docs.godotengine.org/en/stable/contributing/development/compiling/introduction_to_the_buildsystem.html#development-and-production-aliases

This entire PR stems from the idea that many first-time contributors skip the Introduction documentation and go directly to their personal platform's installation instructions. Any ideas on the best way to give them the information they need as unobtrusively as possible?

**Comment by bruvzg** [comment]
> Alternatively, I could just put a link to the Introduction page's development and production aliases headline

This probably is a better option, and will avoid unnecessary duplication of the same info on multiple pages. `use breakpoints while developing` feels too broad and does not mention any negative effects. Most users probably do not want to use it, since disabling optimizations make the engine a lot slower.

**Comment by tetrapod00** [comment]
What about another tip or note box, just before or after the one about the one about `production=yes`?

> Tip
> If you are compiling Godot to make changes or contribute to the engine, you may want to use the SCons options `dev_build=yes` or `dev_mode=yes`. See https://docs.godotengine.org/en/stable/contributing/development/compiling/introduction_to_the_buildsystem.html#development-and-production-aliases for more info.

File: contributing/development/compiling/compiling_for_linuxbsd.rst
Comment (tetrapod00) [inline contributing/development/compiling/compiling_for_linuxbsd.rst:254]:
```suggestion
    See :ref:`doc_introduction_to_the_buildsystem_development_and_production_aliases`
```
We can use internal links here.

You'll also have to add an anchor to the section header in `Introduction to the buildsystem`. Note that there is an underscore `_` in the anchor definition but not the reference.

```diff
- Development and production aliases
- ----------------------------------
+ .. _doc_introduction_to_the_buildsystem_development_and_production_aliases:
+
+ Development and production aliases
+ ----------------------------------
```

The same changes need to be applied to the other two tip boxes.

**Review by tetrapod00 — CHANGES_REQUESTED** [review]
I suggested the current wording for the tip boxes, so I approve this addition. I think it's a net improvement to the pages. It does need one more change to use internal link syntax.

In the future we should probably do something to address the accumulation of tip and note boxes in these pages, but since they are out-of-the-way contributing pages, it seems fine.

**Comment by tetrapod00** [comment]
@jonathansekela Not sure if you're still working on changes, but you still need to add an anchor to the section heading that is being linked to. (The internal link syntax doesn't automatically generate one, we have to add them manually whenever we link from other pages)

**Comment by jonathansekela** [comment]
dear god everyone saw that too

I'm so sorry, that was an abysmal showing on my part. I'm not quite sure what went wrong, but something definitely went wrong when I tried to rebase. I promise I tried to follow the instructions, but obviously not very well. I finally managed to get my fork of godot-docs squashed down to one commit. I'll continue by adding the anchor to the section heading.

**Comment by tetrapod00** [comment]
It's okay, it's not very straightforward the first time you do it. I've messed up my share of rebases too.

Looks like you got to a single squashed commit. FYI now that you have a single commit (and in future PRs), you can use an [amend workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request) as an alternative to multiple commits and squashing.

**Comment by jonathansekela** [comment]
Alright, my fork should be ship-shape and ready for review now. I figured out why my amend didn't work as well - I was syncing, not force-pushing. Please let me know if you see anything else that might be a problem.

**Review by tetrapod00 — APPROVED** [review]
I'm going to give other maintainers a chance to look this over before merging, but this looks good to me 👍

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Merged. Thanks and congrats on your first merged contribution!

**Comment by mhilbrunner** [comment]
Cherrypicked to 4.3 in https://github.com/godotengine/godot-docs/pull/10347.

**Pull request merged**
